query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs

```diff
@@ -348,61 +357,86 @@ impl ChainingHashTable {
+                        Some(v) => {
+                            let probe_result_ptrs = v.get_value();
+                            build_indexs.extend_from_slice(probe_result_ptrs);
+                            for _ in probe_result_ptrs {
+                                probe_indexs.push(i as u32);
+                            }
+                            validity.extend_constant(probe_result_ptrs.len(), true);
                         }
-                    } else {
-                        // No matched row for current probe row
-                        // Create a NULL block for right side
-                        for field in data_schema_fields.iter() {
-                            let nullable_data_type = wrap_nullable(field.data_type());
-                            columns.push(
-                                nullable_data_type.create_constant_column(&DataValue::Null, 1)?,
-                            );
+                        None => {
+                            probe_indexs.push(i as u32);
+                            validity.push(false);
                         }
-                    };
-                    let build_block =
-                        DataBlock::create(self.row_space.data_schema.clone(), columns);
-                    let probe_block = DataBlock::block_take_by_indices(input, &[i as u32])?;
-                    let merged_block = self.merge_block(&build_block, &probe_block)?;
-                    results.push(self.filter_block(&probe_block, merged_block)?);
+                    }
+                }
+                let validity: Bitmap = validity.into();
+                let build_block = self.row_space.gather(&build_indexs)?;
+                let mut nullable_columns = Vec::with_capacity(build_block.num_columns());
+
+                for column in build_block.columns() {
+                    if column.is_null() {
+                        nullable_columns.push(column.clone());
+                    } else if column.is_nullable() {
+                        let col: &NullableColumn = Series::check_get(column)?;
+                        let new_validity = col.ensure_validity() & (&validity);
+                        nullable_columns.push(NullableColumn::wrap_inner(
+                            col.inner().clone(),
+                            Some(new_validity),
+                        ));
+                    } else {
+                        nullable_columns.push(NullableColumn::wrap_inner(
+                            column.clone(),
+                            Some(validity.clone()),
+                        ));
+                    }
                 }
+
+                let nullable_build_block =
+                    DataBlock::create(self.row_space.data_schema.clone(), nullable_columns);
+                let probe_block = DataBlock::block_take_by_indices(input, &probe_indexs)?;
+
+                let merged_block = self.merge_eq_block(&nullable_build_block, &probe_block)?;
```

Comment:
Is the core optimization point for Left outer Join here?

Reply:
Part of, and we avoid building small blocks inside the loop.